File: duna-core/src/main/java/io/duna/core/implementation/MethodCallDelegator.java
/*
 * Copyright (c) 2016 Duna Open Source Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.implementation;

import co.paralleluniverse.fibers.Suspendable;
import net.bytebuddy.ByteBuddy;
import net.bytebuddy.description.annotation.AnnotationDescription;
import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
import net.bytebuddy.implementation.MethodCall;
import net.bytebuddy.implementation.bytecode.assign.Assigner;

import java.lang.reflect.Method;

public interface MethodCallDelegator {

    Object invoke(Object ... args);

    static MethodCallDelegator to(Object delegate, Method method) {
        AnnotationDescription suspendable = AnnotationDescription.Builder
            .ofType(Suspendable.class).build();

        try {
            // @formatter:off
            return (MethodCallDelegator) new ByteBuddy()
                .subclass(Object.class)
                .implement(MethodCallDelegator.class)
                    .intercept(
                    MethodCall
                        .invoke(method)
                        .on(delegate)
                        .withArgumentArrayElements(0, method.getParameterCount())
                        .withAssigner(Assigner.DEFAULT, Assigner.Typing.DYNAMIC)
                    )
                    .annotateMethod(suspendable)
                .make()
                .load(MethodCallDelegator.class.getClassLoader(),
                    ClassLoadingStrategy.Default.INJECTION)
                .getLoaded()
                .newInstance();
            // @formatter:on
        } catch (InstantiationException | IllegalAccessException ex) {
            // TODO fix this
            ex.printStackTrace();
        }

        return null;
    }
}

File: duna-core/src/main/kotlin/io/duna/core/implementation/MethodCallDelegator.kt
/*
 * Copyright (c) 2016 Duna Open Source Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.implementation

import co.paralleluniverse.fibers.Suspendable
import net.bytebuddy.ByteBuddy
import net.bytebuddy.description.ByteCodeElement
import net.bytebuddy.description.annotation.AnnotationDescription
import net.bytebuddy.description.method.MethodDescription
import net.bytebuddy.implementation.MethodCall
import net.bytebuddy.implementation.bytecode.assign.Assigner
import net.bytebuddy.matcher.ElementMatchers.isDeclaredBy
import net.bytebuddy.matcher.ElementMatchers.named
import java.lang.reflect.Method

interface MethodCallDelegator<T> {

  fun invoke(vararg args: Any?): Any?

  companion object {

    @Suppress("UNCHECKED_CAST")
    fun <T> to(delegate: T, method: Method): MethodCallDelegator<T> {
      val suspendableAnnotation = AnnotationDescription.Builder.ofType(Suspendable::class.java).build()

      // @formatter:off
      return ByteBuddy()
        .subclass<Any>(Any::class.java)
        .implement(MethodCallDelegator::class.java)
          .intercept(
            MethodCall
              .invoke(method)
              .on(delegate)
              .withArgumentArrayElements(0, method.parameterCount)
              .withAssigner(Assigner.DEFAULT, Assigner.Typing.DYNAMIC)
          )
          .annotateMethod(suspendableAnnotation)
        .make()
        .load(MethodCallDelegator::class.java.classLoader)
        .loaded
        .newInstance() as MethodCallDelegator<T>
      // @formatter:on
    }
  }
}
